server/src/main/java/org/elasticsearch/index/mapper/ParsedDocument.java

```diff
@@ -35,6 +38,37 @@
 
     private Mapping dynamicMappingsUpdate;
 
+    public static ParsedDocument noopTombstone(String index, String reason) {
+        ParsedDocument doc = deleteTombstone(index, null);
+        // Store the reason of a noop as a raw string in the _source field
+        final BytesRef byteRef = new BytesRef(reason);
+        doc.rootDoc().add(new StoredField(SourceFieldMapper.NAME, byteRef.bytes, byteRef.offset, byteRef.length));
+        return doc;
+    }
+
+    public static ParsedDocument deleteTombstone(String index, String id) {
+        SourceToParse source = new SourceToParse(index, id == null ? "" : id, new BytesArray("{}"), XContentType.JSON);
+        ParseContext.InternalParseContext context = new ParseContext.InternalParseContext(null, 1,null, source, null);
+        if (id != null) {
+            IdFieldMapper.NO_FIELDDATA_INSTANCE.preParse(context);
+        }
+        VersionFieldMapper.INSTANCE.preParse(context);
+        SeqNoFieldMapper.INSTANCE.preParse(context);
```

Comment:
why not keeping an array of these (actually two arrays) and using the existing internalParseDocument? that may clean up also the need for handling null id

Reply:
`internalParseDocument` needs a RootObjectMapper and we really don't need one here.

Reply:
I see. I would look into create an empty root and pass it in, instead of having a lighter variant of internalParseDocument that's specific for the tombstone document. If you index a doc, you should always go through the same checks and code path? I am particularly worried about subsequent changes that may be bring the two out of sync accidentally